world/wguild_mgr.cpp

```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+			g.second->tribute.time_remaining = RuleI(Guild, TributeTime);
+			g.second->tribute.timer.Start(RuleI(Guild, TributeTime));
+
+			guild_mgr.DBSetGuildFavor(g.first, g.second->tribute.favor);
+			guild_mgr.DBSetTributeTimeRemaining(g.first, RuleI(Guild, TributeTime));
+
+			LogGuilds("Timer reset.  Do tribute work - Points, send timer restart to zones, etc.");
+
+			SendGuildTributeFavorAndTimer(g.first, g.second->tribute.favor, g.second->tribute.timer.GetRemainingTime());
+			//ServerPacket* sp = new ServerPacket(ServerOP_GuildTributeFavAndTimer, sizeof(GuildTributeFavorTimer_Struct));
+			//GuildTributeFavorTimer_Struct* data = (GuildTributeFavorTimer_Struct*)sp->pBuffer;
+			//data->guild_id = g.first;
+			//data->guild_favor = g.second->tribute.favor;
+			//data->tribute_timer = g.second->tribute.time_remaining;
+			//data->trophy_timer = 0;
+			//
+			//for (auto const& z : zoneserver_list.getZoneServerList()) {
+			//	auto r = z.get();
+			//	if (r->GetZoneID() > 0) {
+			//		r->SendPacket(sp);
+			//	}
+			//}
+			//safe_delete(sp);
+
+		}
+		else if (g.second->tribute.send_timer && 
+			((g.second->tribute.timer.GetRemainingTime()/1000) % (RuleI(Guild, TributeTimeRefreshInterval) /1000)) == 0 &&
+			!g.second->tribute.timer.Check()
+			){
+			g.second->tribute.send_timer = false;
+			g.second->tribute.time_remaining = g.second->tribute.timer.GetRemainingTime();
+			SendGuildTributeFavorAndTimer(g.first, g.second->tribute.favor, g.second->tribute.time_remaining);
+			guild_mgr.DBSetTributeTimeRemaining(g.first, g.second->tribute.time_remaining);
+			LogGuilds("Timer Frequency [{}] ms hit\.  Sending time [{}] to guild clients\.",
+				RuleI(Guild, TributeTimeRefreshInterval),
+				g.second->tribute.time_remaining
+			);
+		}
+		else if (!g.second->tribute.send_timer &&
+			((g.second->tribute.timer.GetRemainingTime() / 1000) % (RuleI(Guild, TributeTimeRefreshInterval) / 1000)) > 0 &&
+			!g.second->tribute.timer.Check()
+			) {
+			g.second->tribute.send_timer = true;
+		}
+	}
+}
+
+uint32 WorldGuildManager::GetGuildTributeCost(uint32 guild_id)
+{
+	auto guild_members = client_list.GetGuildClientsWithTributeOptIn(guild_id);
+	auto total = guild_members.size();
+	auto total_cost = 0;
+
+	auto guild = guild_mgr.GetGuildByGuildID(guild_id);
+	if (guild) {
+
+		TributeData& d1 = tribute_list[guild->tribute.id_1];
+		TributeData& d2 = tribute_list[guild->tribute.id_2];
+
+		uint32 cost_id1 = d1.tiers[guild->tribute.id_1_tier].cost;
+		uint32 cost_id2 = d2.tiers[guild->tribute.id_2_tier].cost;
+		uint32 level_id1 = d2.tiers[guild->tribute.id_1_tier].level;
+		uint32 level_id2 = d2.tiers[guild->tribute.id_2_tier].level;
+
+		for (auto const& m : guild_members) {
+			if (m.second->level() >= level_id1) {
+				total_cost += cost_id1;
+			}
+			if (m.second->level() >= level_id2) {
+				total_cost += cost_id2;
+			}
+		}
+	}
+	return total_cost;
+}
+
+bool WorldGuildManager::LoadTributes() {
```

Comment:
Put together a compiling example of what this function can look like with repository calls. Code is simpler and easier to read. It does need to be fully tested.

```cpp
bool WorldGuildManager::LoadTributes()
{
	TributeData td;
	memset(&td.tiers, 0, sizeof(td.tiers));
	td.tier_count = 0;

	tribute_list.clear();

	auto tributes = TributesRepository::All(*m_db);

	for (auto const &t: tributes) {
		td.name        = t.name;
		td.description = t.descr;
		td.unknown     = t.unknown;
		td.is_guild    = t.isguild == 0 ? false : true;

		tribute_list[t.id] = td;
	}

	LogInfo("Loaded [{}] tributes", Strings::Commify(tributes.size()));

	auto tribute_levels = TributeLevelsRepository::GetWhere(*m_db, "TRUE ORDER BY tribute_id, level");

	for (auto const &t: tribute_levels) {
		uint32 id = t.tribute_id;

		if (tribute_list.count(id) != 1) {
			LogError("Error in LoadTributes: unknown tribute [{}] in tribute_levels", (unsigned long) id);
			continue;
		}

		TributeData &cur = tribute_list[id];
		if (cur.tier_count >= MAX_TRIBUTE_TIERS) {
			LogError("Error in LoadTributes: on tribute [{}] more tiers defined than permitted", (unsigned long) id);
			continue;
		}

		TributeLevel_Struct &s = cur.tiers[cur.tier_count];

		s.level           = t.level;
		s.cost            = t.cost;
		s.tribute_item_id = t.item_id;

		cur.tier_count++;
	}

	LogInfo("Loaded [{}] tribute levels", Strings::Commify(tribute_levels.size()));

	return true;
}
```

Reply:
Updated for next commit